File: 2324A/NewBB mocks/mock2/pairingsocks_Amanda.cpp
/*
*OpenKattis
*pairingsocks
*Code by: Amanda Lim
*/

#include <bits/stdc++.h>
using namespace std;

int main ()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    
    int N;
    cin >> N;

    queue <int> sockpile;
    for (int n = 0; n < 2*N; n++) {
        int sock;
        cin >> sock;
        sockpile.push(sock);
    }

    stack <int> othersockpile;
    int moves = 0;
    while (!sockpile.empty()) {
        bool taken = false;

        int sock = sockpile.front();
        sockpile.pop();
      
        if (!othersockpile.empty()) {
            if (othersockpile.top() == sock) {
                othersockpile.pop();
                moves++;
                taken = true;
            }
        }

        if (!taken) {
            othersockpile.push(sock);
            moves++;
        }
    }

    if (othersockpile.empty())
        cout << moves << endl;
    else
        cout << "impossible" << endl;

    return 0;
}

File: 2324A/NewBB mocks/mock2/pairingsocks_Amanda.py
'''
OpenKattis
pairingsocks
Code by: Amanda Lim
'''

from queue import LifoQueue, Queue
N = int(input())
sockpile = Queue(maxsize = 2*N)

socks = input().split()
for i in socks:
    sockpile.put(i)

othersockpile = LifoQueue(maxsize = 2*N)
moves = 0

while (sockpile.qsize() > 0):
    sock = sockpile.get_nowait()
    taken = False

    if (othersockpile.qsize() > 0):
        othersock = othersockpile.get_nowait()
        if (sock == othersock):
            moves += 1
            taken = True
        else:
            othersockpile.put_nowait(othersock)
    
    if (not taken):
        othersockpile.put_nowait(sock)
        moves += 1

if (othersockpile.qsize() == 0):
    print(moves)
else:
    print("impossible")